docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md

```diff
@@ -0,0 +1,274 @@
+reference counter, and each shallow volume deletion would decrement it.
+Calling `DeleteSnapshot` would decrement the reference counter as well.
+
+The subvolume snapshot would be removed from the Ceph cluster only once the
+reference count reaches zero. Note that this behavior would permit calling
+`DeleteSnapshot` even if it is still referenced by shallow volumes.
+
+* `DeleteSnapshot`:
+  * `refcount == 0` or the respective _refcount_ RADOS object doesn't exist:
+    delete the backing snapshot too.
+  * `refcount > 0`: keep the backing snapshot.
+* `DeleteVolume`:
+  * `refcount == 0`: delete the backing snapshot too.
+  * `refcount > 0`: keep the backing snapshot.
+
+To enable creating shallow volumes from snapshots that were provisioned by
+older versions of cephfs-csi (i.e. before this feature is introduced),
+`CreateVolume` for shallow volumes would also create the _refcount_ object in
+case it's missing. It would be initialized to two: the source snapshot and the
+newly created shallow volume.
+
+##### Concurrent access to _refcount_ objects
+
+We can't do operations on integers in RADOS objects atomically, which means we
+need to do all synchronization ourselves when modifying the value of the counter.
+
+From a single master perspective, concurrent access to _refcount_ can be
+protected by mutex locks and is easy to implement. `cephfs.ControllerServer`
+already offers `VolumeLocks`, `SnapshotLocks` and `OperationLocks`.
+
+Mutexes won't cover situations where multiple masters may be active though
+(during/after leader election), and modifying _refcount_ would still race in
+this case. Misbehaving master that lost leadership would still be able to
+access resources on the Ceph cluster. Racing on _refcount_ could then cause
+leaking the _refcount_ object itself and/or the subvolume snapshot, or the
+provisioned shallow volume could point to a subvolume snapshot that's already
+been deleted. This is not easily fixable without the individual master replicas
+being aware of the leader election process and is out-of-scope for this
+proposal.
+
+#### `CreateVolume`
+
+A read-only volume with snapshot source would be created under these conditions:
+
+1. `CreateVolumeRequest.volume_content_source` is a snapshot,
+1. `CreateVolumeRequest.volume_capabilities[*].access_mode` is any of read-only
+   volume access modes.
+1. Possibly other volume parameters in `CreateVolumeRequest.parameters`
+   specific to shallow volumes.
+
+`CreateVolumeResponse.Volume.volume_context` would then contain necessary
+information to identify the source subvolume / snapshot.
+
+Things to look out for:
+
+* _What's the volume size?_
+
+  It doesn't consume any space on the filesystem. `Volume.capacity_bytes` is
+  allowed to contain zero. We could use that.
+* _What should be the requested size when creating the volume (specified e.g.
+  in PVC)?_
+
+  This one is tricky. CSI spec allows for
```

Comment:
Yes even though there is no consumption we need to return the requested size in response to make sure no Error happens at the kubernetes level.